klib/syslog.c

```diff
@@ -0,0 +1,338 @@
+} kfuncs;
+
+closure_function(3, 1, void, syslog_file_write_complete,
+                 void *, buf, u64, len, sg_list, sg,
+                 status, s)
+{
+    if (is_ok(s)) {
+        fetch_and_add(&syslog.file_offset, bound(len));
+        if (syslog.file_offset >= SYSLOG_FILE_MAXSIZE) {
+            /* Do log rotation: rename the current log file by adding a ".1" extension to the file
+             * name, and continue logging on a new file. */
+            buffer archive_file = stack_allocate(sizeof(struct buffer));
+            bytes file_path_len = buffer_length(syslog.output_file);
+            archive_file->contents = stack_allocate(file_path_len + 2);
+            archive_file->start = 0;
+            kfuncs.runtime_memcpy(archive_file->contents, buffer_ref(syslog.output_file, 0),
+                file_path_len);
+            byte(archive_file, file_path_len) = '.';
+            byte(archive_file, file_path_len + 1) = '1';
+            archive_file->end = archive_file->length = file_path_len + 2;
+            boolean unlock = spin_try(&syslog.lock);
+            sysreturn ret = kfuncs.fs_rename(syslog.output_file, archive_file);
+            if (ret == 0) {
+                fsfile file = kfuncs.fsfile_open_or_create(syslog.output_file);
+                syslog.file_offset = 0;
+                syslog.fs_write = kfuncs.fsfile_get_writer(file);
+            } else {
+                syslog.fs_write = 0;    /* stop logging */
+            }
+            if (unlock)
+                spin_unlock(&syslog.lock);
+        }
+    } else {
+        kfuncs.timm_dealloc(s);
+    }
+    deallocate(syslog.h, bound(buf), SYSLOG_BUF_LEN);
+    kfuncs.deallocate_sg_list(bound(sg));
+    closure_finish();
+}
+
+static void syslog_file_flush(void)
+{
+    syslog_lock();
+    sg_list sg = syslog.file_sg;
+    if (sg) {
+        syslog.file_sg = 0;
+        u64 len = syslog.file_sgb->offset;
+        status_handler sh =
+                closure(syslog.h, syslog_file_write_complete, syslog.file_sgb->buf, len, sg);
+        if (sh != INVALID_ADDRESS) {
+            syslog.file_sgb->offset = 0;
+            apply(syslog.fs_write, sg, irangel(syslog.file_offset, len), sh);
+        } else {
+            /* Discard logged data. */
+            deallocate(syslog.h, syslog.file_sgb->buf, SYSLOG_BUF_LEN);
+            kfuncs.deallocate_sg_list(sg);
+        }
+    }
+    syslog_unlock();
+}
+
+define_closure_function(0, 1, void, syslog_timer_func,
+                        u64, overruns)
+{
+    syslog.file_flush_timer = INVALID_ADDRESS;
+    syslog_file_flush();
+}
+
+static void syslog_file_write(void *d, const char *s, bytes count)
+{
+    if (!syslog.fs_write || (count > SYSLOG_BUF_LEN))
+        return;
+    if (!syslog.file_sg) {
+        syslog.file_sg = kfuncs.allocate_sg_list();
+        if (syslog.file_sg == INVALID_ADDRESS)
+            return;
+        syslog.file_sgb = kfuncs.sg_list_tail_add(syslog.file_sg, SYSLOG_BUF_LEN);
```

Comment:
This isn't critical, and if it even makes sense to do it could be done in another PR, but I would suggest doing this kind of buffering on the front end of the syslog console write. There are a few reasons for this. There is quite a lot of work to do for every console write, and In situations where there are frequent and especially short writes, this overhead could add up and have a performance impact. With buffering on the front end, the actual file and/or udp write could be deferred somewhat by a timer, reducing the latency of the console writes and amortizing the cost of the write operation over a larger amount of output. It would also reduce the number of file writes / udp frames sent for a given amount of output if the output is bursty. Also, allowance for a backlog (up to a configurable limit) would allow the capture of output that occurs before the default interface is up and name resolution has finished.

Reply:
The syslog protocol requires each UDP packet to convey exactly one message, so we can't group different messages in a single packet (well, we could do it, but I assume most users would expect each syslog message to correspond to a printf() or similar function call in their program, especially if they are interested in the timestamp associated to each message). So I think buffering shouldn't be done for UDP.

Reply:
I see. I guess the only real advantages to buffering, then, (while still sending one packet per queued message) would be to gather the UDP writes together as a deferred operation rather than going through the whole stack on each console output, thereby reducing the latency of writes, and also to be able to maintain some degree of backlog to capture early messages before the server becomes reachable.